packages/discovery-react-components/src/components/DiscoverySearch/DiscoverySearch.tsx

```diff
@@ -480,14 +484,46 @@ const DiscoverySearch: FC<DiscoverySearchProps> = ({
   );
 
   const handleFetchDocuments = useCallback(
-    (filterString: string, searchResponse: DiscoveryV2.QueryResponse | null): void => {
-      fetchDocuments(filterString, filteredResponse => {
+    (
+      filterString: string,
+      collections?: string[],
+      searchResponse?: DiscoveryV2.QueryResponse
+    ): void => {
+      fetchDocuments(filterString, collections || [], filteredResponse => {
         if (searchResponse && searchResponse.results && filteredResponse.results) {
+          const results = [...searchResponse.results];
+          const filteredResults = [...filteredResponse.results];
+
+          // replace existing doc data in `searchResponse` with new filtered response data
+          while (filteredResults.length) {
+            const filteredResult = filteredResults.pop();
+            if (!filteredResult) {
+              continue;
+            }
+
+            const idx = results.findIndex(
+              res =>
+                res.document_id === filteredResult?.document_id &&
+                res.result_metadata.collection_id === filteredResult.result_metadata.collection_id
+            );
+
+            if (idx !== -1) {
+              // merge results, since filteredResult may not contain all fields from a previous
+              // query (like passages)
+              const mergedResult = {
+                ...results[idx],
+                ...filteredResult
+              };
+              results.splice(idx, 1, mergedResult);
+            } else {
+              results.push(filteredResult);
+            }
+          }
+
           setSearchResponse({
             ...searchResponse,
-            matching_results:
-              (searchResponse.matching_results as number) + filteredResponse.results.length,
-            results: [...searchResponse.results, ...filteredResponse.results]
+            matching_results: results.length,
+            results
```

Comment:
The old code simply appended the new doc data to the existing `searchResponseStore` data. But this lead to duplicated documents, as well as more items than were there previously.

The code above instead will merge the document data with any existing data, assuming that the doc and collection IDs match. Any unmatched docs are still appended.

This could be considered a BREAKING CHANGE, but I think this is actually fixing a bug in the previous implementation.